src/util/mapbox.js

```diff
@@ -119,3 +126,108 @@ function formatUrl(obj: UrlObject): string {
     const params = obj.params.length ? `?${obj.params.join('&')}` : '';
     return `${obj.protocol}://${obj.authority}${obj.path}${params}`;
 }
+
+class TurnstileEvent {
+    STORAGE_TOKEN: string;
+    localStorageAvailable: boolean;
+    eventData: Object;
+    queue: Array<number>;
+    pending: boolean
+    pendingRequest: ?Cancelable;
+
+    constructor() {
+        this.STORAGE_TOKEN = 'mapbox.turnstileEventData';
+        this.localStorageAvailable = storageAvailable('localStorage');
+        this.eventData = { anonId: null, lastSuccess: null };
+        this.queue = [];
+        this.pending = false;
+        this.pendingRequest = null;
+    }
+
+    queueRequest(date: number) {
+        this.queue.push(date);
+        this.processRequests();
+    }
+
+    processRequests() {
+        if (this.pending || this.queue.length === 0) {
+            return;
+        }
+        this.pending = true;
+        let dueForEvent = false;
+        if (!this.eventData.anonId || !this.eventData.lastSuccess) {
+            //Retrieve cached data
+            if (this.localStorageAvailable) {
+                try {
+                    const data = window.localStorage.getItem(this.STORAGE_TOKEN);
+                    if (data) {
+                        const json = JSON.parse(data);
+                        this.eventData.anonId = json.anonId;
+                        this.eventData.lastSuccess = Number(json.lastSuccess);
+                    }
+                } catch (e) {
+                    warnOnce('Unable to read from LocalStorage');
+                }
+            }
+        }
+
+        if (!validateUuid(this.eventData.anonId)) {
+            this.eventData.anonId = uuid();
+            dueForEvent = true;
+        }
+        const nextUpdate = this.queue.shift();
+
+        // Record turnstile event once per calendar day.
+        if (this.eventData.lastSuccess) {
```

Comment:
What's the expected behavior if localStorage is not available?  As of now, it looks like we'd only send a turnstile event once per page in that case, since `lastSuccess` would never be set.

Reply:
`eventData` caches the id and success values within a single session. But I see the problem, this value needs to be updated irrespective of `localStorageAvailable` when a request succeeds.

If `localStorage` is not available, there may be more than one event sent per calendar day, but the events server de-duplicates those.